File: tests/python_tests/agg_rasterizer_integer_overflow_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from nose.tools import eq_
from utilities import run_all
import mapnik
import json

# geojson box of the world
geojson  = { "type": "Feature", "properties": { }, "geometry": { "type": "Polygon", "coordinates": [ [ [ -17963313.143242701888084, -6300857.11560364998877 ], [ -17963313.143242701888084, 13071343.332991421222687 ], [ 7396658.353099936619401, 13071343.332991421222687 ], [ 7396658.353099936619401, -6300857.11560364998877 ], [ -17963313.143242701888084, -6300857.11560364998877 ] ] ] } }

def test_that_coordinates_do_not_overflow_and_polygon_is_rendered_memory():
  expected_color = mapnik.Color('white')
  projection = '+init=epsg:4326'
  ds = mapnik.MemoryDatasource()
  context = mapnik.Context()
  feat = mapnik.Feature.from_geojson(json.dumps(geojson),context)
  ds.add_feature(feat)
  s = mapnik.Style()
  r = mapnik.Rule()
  sym = mapnik.PolygonSymbolizer()
  sym.fill = expected_color
  r.symbols.append(sym)
  s.rules.append(r)
  lyr = mapnik.Layer('Layer',projection)
  lyr.datasource = ds
  lyr.styles.append('style')
  m = mapnik.Map(256,256,projection)
  m.background_color = mapnik.Color('green')
  m.append_style('style',s)
  m.layers.append(lyr)
  # 17/20864/45265.png
  m.zoom_to_box(mapnik.Box2d(-13658379.710221574,6197514.253362091,-13657768.213995293,6198125.749588372))
  # works 15/5216/11316.png
  #m.zoom_to_box(mapnik.Box2d(-13658379.710221574,6195679.764683247,-13655933.72531645,6198125.749588372))
  im = mapnik.Image(256,256)
  mapnik.render(m,im)
  eq_(im.get_pixel(128,128),expected_color.packed())

def test_that_coordinates_do_not_overflow_and_polygon_is_rendered_csv():
  expected_color = mapnik.Color('white')
  projection = '+init=epsg:4326'
  ds = mapnik.MemoryDatasource()
  context = mapnik.Context()
  feat = mapnik.Feature.from_geojson(json.dumps(geojson),context)
  ds.add_feature(feat)
  geojson_string = "geojson\n'%s'" % json.dumps(geojson['geometry'])
  ds = mapnik.Datasource(**{'type':'csv','inline':geojson_string})
  s = mapnik.Style()
  r = mapnik.Rule()
  sym = mapnik.PolygonSymbolizer()
  sym.fill = expected_color
  r.symbols.append(sym)
  s.rules.append(r)
  lyr = mapnik.Layer('Layer',projection)
  lyr.datasource = ds
  lyr.styles.append('style')
  m = mapnik.Map(256,256,projection)
  m.background_color = mapnik.Color('green')
  m.append_style('style',s)
  m.layers.append(lyr)
  # 17/20864/45265.png
  m.zoom_to_box(mapnik.Box2d(-13658379.710221574,6197514.253362091,-13657768.213995293,6198125.749588372))
  # works 15/5216/11316.png
  #m.zoom_to_box(mapnik.Box2d(-13658379.710221574,6195679.764683247,-13655933.72531645,6198125.749588372))
  im = mapnik.Image(256,256)
  mapnik.render(m,im)
  eq_(im.get_pixel(128,128),expected_color.packed())

if __name__ == "__main__":
    exit(run_all(eval(x) for x in dir() if x.startswith("test_")))

File: test/standalone/agg_rasterizer_integer_overflow_test.cpp
#define CATCH_CONFIG_MAIN
#include "catch.hpp"

#include <mapnik/map.hpp>
#include <mapnik/memory_datasource.hpp>
#include <mapnik/json/feature_parser.hpp>
#include <mapnik/feature_type_style.hpp>
#include <mapnik/rule.hpp>
#include <mapnik/symbolizer.hpp>
#include <mapnik/layer.hpp>
#include <mapnik/agg_renderer.hpp>
#include <mapnik/debug.hpp>

#include <iostream>

// geojson box of the world
const std::string
  geojson("{ \"type\": \"Feature\", \"properties\": { }, \"geometry\": { \"type\": \"Polygon\", \"coordinates\": [ [ [ "
          "-17963313.143242701888084, -6300857.11560364998877 ], [ -17963313.143242701888084, 13071343.332991421222687 "
          "], [ 7396658.353099936619401, 13071343.332991421222687 ], [ 7396658.353099936619401, "
          "-6300857.11560364998877 ], [ -17963313.143242701888084, -6300857.11560364998877 ] ] ] } }");

TEST_CASE("agg_rasterizer_integer_overflow")
{
    SECTION("coordinates_do_not_overflow_and_polygon_is_rendered")
    {
        auto expected_color = mapnik::color("white");

        mapnik::Map m(256, 256);
        m.set_background(mapnik::color("black"));

        mapnik::feature_type_style s;
        {
            mapnik::rule r;
            mapnik::polygon_symbolizer sym;
            mapnik::put(sym, mapnik::keys::fill, expected_color);
            mapnik::put(sym, mapnik::keys::clip, false);
            r.append(std::move(sym));
            s.add_rule(std::move(r));
        }
        m.insert_style("style", std::move(s));

        mapnik::layer lyr("Layer");
        lyr.styles().emplace_back("style");
        {
            auto ds = std::make_shared<mapnik::memory_datasource>(mapnik::parameters());
            auto context = std::make_shared<mapnik::context_type>();
            auto f = std::make_shared<mapnik::feature_impl>(context, 0);
            REQUIRE(mapnik::json::from_geojson(geojson, *f));
            ds->push(f);
            lyr.set_datasource(ds);
        }
        m.add_layer(std::move(lyr));

        // 17/20864/45265.png
        m.zoom_to_box(
          mapnik::box2d<double>(-13658379.710221574, 6197514.253362091, -13657768.213995293, 6198125.749588372));

        // works 15/5216/11316.png
        // m.zoom_to_box(mapnik::box2d<double>(-13658379.710221574,6195679.764683247,-13655933.72531645,6198125.749588372));

        mapnik::image_rgba8 im(256, 256);
        {
            mapnik::agg_renderer<mapnik::image_rgba8> ren(m, im);
            ren.apply();
        }

        REQUIRE(im(128, 128) == expected_color.rgba());
    } // SECTION
} // TEST_CASE
